File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:


        wordDict = set(wordDict) # Put all words in a set which makes it O(1) retrieval time
        dp = [False] * (len(s) + 1)
        dp[-1] = True

        for i in range(len(s) - 1, -1, -1):
            found = False
            for j in range(i, len(s)):
                pot_word = s[i:j + 1]
                if pot_word in wordDict and dp[i + len(pot_word)]:
                    found = True
                    break
            dp[i] = found
        
        return dp[0]


        

File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> mp;
        for (auto word: wordDict) mp.insert(word);

        int n = s.size();

        vector<bool> dp(n + 1, false);
        dp[n] = true;

        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                string sub = s.substr(i, j - i + 1);
                if (mp.find(sub) != mp.end()) {
                    int l = sub.size();
                    dp[i] = dp[i + l];
                    if (dp[i]) break;
                }
            }
        }
        return dp[0];
    }
};